### PR TITLE
Added article and solution for GenerateParantheses

### PR DESCRIPTION
**Pull request by Arihant1467**
Added Article and Solution for [Genrate Parantheses](https://leetcode.com/problems/generate-parentheses/)

### CODE REVIEW

File: Backtracking/GenerateParantheses/README.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+<img align="center" alt="Question Screenshot" src="./image-1.png">
+</p>
+
+
+---
+
+### Solution : Backtracking
+
+####  Motivation
+So we are given `N` pairs of parentheses and return all the valid expressions. The above picture has all valid expressions for `n=3`.
```
Comment (edorado93) [inline Backtracking/GenerateParantheses/README.md]: 1. Maybe rephrase the first sentence as `So we are given N pairs of parentheses and we want to return all the valid expressions possible by using those parentheses`.

2. The picture `below`

File: Backtracking/GenerateParantheses/README.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+Below is the recursion tree formed by backtracking for `n=2`.
+
+<p>
+<img align="center" alt="Question Screenshot" src="./../../Backtracking/GenerateParantheses/complete-recursion-tree.png">>
+</p>
+
+So, from the above diagram if we traverse all the paths from root to leaf nodes we will get all the possible expressions for `n=2`. All expressions are listed as below: <br>
+
+` (((( `
```
Comment (edorado93) [inline Backtracking/GenerateParantheses/README.md]: I think you want each expression on a new line.  Instead of using a single  `  here, use  ```  for a code block and put each expression on a single line.

File: Backtracking/GenerateParantheses/README.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+` )((( ` 
+` )(() ` 
+` )()( ` 
+` )()) ` <br>
+` ))(( ` 
+` ))() ` 
+` )))( ` 
+` )))) ` 
+
+After a careful observation of the tree and the expressions, all these expressions can be formed by making a <b>Postorder Traversal</b> of the tree. Out of these expressions we only add the valid one into our result.
```
Comment (edorado93) [inline Backtracking/GenerateParantheses/README.md]: `After careful observation of the tree, we can see that all these expressions ...`

File: Backtracking/GenerateParantheses/README.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+####  Motivation
+So we are given `N` pairs of parentheses and return all the valid expressions. The above picture has all valid expressions for `n=3`.  
+
+The length of any expression will be equal to `n*2`. So, for every index in the expression we have 2 ways to fill the index either `(` or `)`. So  we try to recursively fill every index of the the expression either by using `(` or `)` and check the validity of the expression. This can be accomplished by the help of Backtracking. 
+
+Below is the recursion tree formed by backtracking for `n=2`.
+
+<p>
+<img align="center" alt="Question Screenshot" src="./../../Backtracking/GenerateParantheses/complete-recursion-tree.png">>
```
Comment (edorado93) [inline Backtracking/GenerateParantheses/README.md]: If it's not too much of a hassle, can you redraw the figure using round brackets rather than curly?

File: Backtracking/GenerateParantheses/README.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+` ))(( ` 
+` ))() ` 
+` )))( ` 
+` )))) ` 
+
+After a careful observation of the tree and the expressions, all these expressions can be formed by making a <b>Postorder Traversal</b> of the tree. Out of these expressions we only add the valid one into our result.
+
+
+#### Algorithm
+1. We start with an empty string  and mark it as `level=0`.
```
Comment (edorado93) [inline Backtracking/GenerateParantheses/README.md]: leave spaces between expressions e.g.

`level = 0`

File: Backtracking/GenerateParantheses/README.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+` )))( ` 
+` )))) ` 
+
+After a careful observation of the tree and the expressions, all these expressions can be formed by making a <b>Postorder Traversal</b> of the tree. Out of these expressions we only add the valid one into our result.
+
+
+#### Algorithm
+1. We start with an empty string  and mark it as `level=0`. 
+2. We then make a postorder traversal of the tree till `level=n*2` by keeping the track of the expression which is being formed by the traversal. 
+3. At `level=n*2` we check if the formed expression is a valid one. If its valid then we add it otherwise we discard it.
```
Comment (edorado93) [inline Backtracking/GenerateParantheses/README.md]: This operation takes `O(N)` time as well. So the time complexity of your solution is actually `O(N * 2^N - 1)`. There's a better way of doing this, you don't need to check validity in the end. 

Essentially, you keep a track of a `left_parentheses_count` and `right_parentheses_count` and an index. The base condition is when `index` reaches `N`. As you mentioned, at every index you have two choices. You can either add a left parentheses and in the next recursive call pass `left_parentheses_count + 1`. 

For adding the right parentheses, you need to check first `right_parentheses_count < left_parentheses_count` because if this is not the case, then you would get an invalid string. So, only  if this condition is satisfied, you recurse further with `right_parentheses_count + 1`. When index reaches N, you simply check if `left_parentheses_count == right_parentheses_count`. Then the approach becomes `O(2^N)`. 

I'd recommend, you modify the existing approach and add this information in the algorithm itself.

File: Backtracking/GenerateParantheses/solution.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        # We will call this function in a recursive manner by adding both the
+        # opening and closing bracket to the current expression which will check 
+        # the current path is a valid path from root to leaf node.
+        self.helper(expression+"(",result,level+1,n)
+        self.helper(expression+")",result,level+1,n)
+        
+    
+    
+    # This function checks the validity of the expression
+    def checker(self,expression):
```
Comment (edorado93) [inline Backtracking/GenerateParantheses/solution.py]: As mentioned in my comment before, you can get rid of this and solve it in `O(1)` time by using two counters.

File: Backtracking/GenerateParantheses/README.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+#### Complexity Analysis
+* Time Complexity: `O(2^n-1)` where n is the number of pairs of parantheses.
```
Comment (edorado93) [inline Backtracking/GenerateParantheses/README.md]: You need to add the space complexity as well.

File: Backtracking/GenerateParantheses/README.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+After a careful observation of the tree and the expressions, all these expressions can be formed by making a <b>Postorder Traversal</b> of the tree. Out of these expressions we only add the valid one into our result.
+
+
+#### Algorithm
+1. We start with an empty string  and mark it as `level=0`. 
+2. We then make a postorder traversal of the tree till `level=n*2` by keeping the track of the expression which is being formed by the traversal. 
+3. At `level=n*2` we check if the formed expression is a valid one. If its valid then we add it otherwise we discard it.
+
+#### Complexity Analysis
```
Comment (edorado93) [inline Backtracking/GenerateParantheses/README.md]: Can you please capitalize N?

File: Backtracking/GenerateParantheses/README.md (hunk 4 above)
Comment (edorado93) [reply Backtracking/GenerateParantheses/README.md]: Also, in your diagram, can you write the final expressions along with the leaf nodes and maybe highlight the valid ones?

File: README.md [hunk 10]
```diff
@@ -405,4 +411,13 @@ There are multiple ways in which you can contribute. There are no "prerequisites
     <td>LeetCode</td>
     <td><a href="https://github.com/DivyaGodayal/CoderChef-Kitchen/tree/master/Strings/Reorganize_String">Link</a></td>
   </tr>
+  <tr>
```
Comment (edorado93) [inline README.md]: We have all our backtracking problems in the `Adhoc` section for now, Can you move yours there as well. In there future, we will have a separate section for `Recursion and Backtracking`. For now just move it there.

**Review by edorado93 — CHANGES_REQUESTED** [review]

**Comment by edorado93** [comment]
Hey, @Arihant1467 lemme know once you're done addressing the `review` related changes. I'll verify things once more and proceed with merging. Thanks for the article!

**Comment by Arihant1467** [comment]
Hi, @edorado93  I have made the requested changes. Let me know if any errors are there.

**Comment by edorado93** [comment]
Hey @Arihant1467, the changes look good to me. I made some minor changes to the README. Also, I moved the images to the respective folder in the `Images` directory. Rest looks good to me. Good job with the diagrams! Welcome to the kitchen 🍩 

template_image_3
